sql: add SwallowSettingOverrideErr session var

msbutler opened the pull request:
When set, an app tenant that runs SET CLUSTER SETTING on a setting overrode by
the system tenant will noop instead of return an error.

Epic: none

Release note: none

msbutler committed:
sql: add SwallowSessionOverrideErr session var

When set, an app tenant that runs SET CLUSTER SETTING on a setting overrode by
the system tenant will noop instead of return an error.

Epic: none

Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/144404)
<!-- Reviewable:end -->

msbutler committed:
upgrades: swallow setting override error in diagnostics migration

Previously if a tenant ran the optInToDiagnosticsStatReporting migration, which
sets diagnostics.reporting.enabled, after the system tenant already overrode
this setting, the migration would enter a fail loop. With this patch, the
migration instead noops.

Epic: none

Release note: none

stevendanna reviewed (approved):
Overall this looks reasonable to me. I suppose on alternative to the early exit would be to allow the cluster setting to be set inside the tenant still.  I don't have a strong opinion on which behaviour is better and since this is only used on the internal executor, we can change it in the future without too much hassle.

rafiss reviewed (approved):
lgtm! just had some suggested nits

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144404#-:-OO2wHZ7EVAgtIQv7ReZ:bwohe8i)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @jeffswenson)

___
*[`pkg/sql/sessiondata/internal.go` line 92 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/144404#-OO2v_9cCtmK786qa-a4:-OO2v_9cCtmK786qa-a5:b456gvy) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a3e56ae1c4deec1097a299ab062ab9630912c49/pkg/sql/sessiondata/internal.go#L92)):*
> ```Go
> 	// SwallowSettingOverrideErr, if set, indicates that when an app tenant
> 	// attempts to set a cluster setting overrode by the system tenant, the query
> 	// no-ops instead of errors.
> ```

nit: it could be worth mentioning that this setting is only used internally, and is not exposed to any `SET`/`SHOW` commands
___
*[`pkg/sql/sessiondata/internal.go` line 93 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/144404#-OO2w8FU01DGBS30ESeH:-OO2w8FU01DGBS30ESeI:b6nlwr8) ([raw file](https://github.com/cockroachdb/cockroach/blob/2a3e56ae1c4deec1097a299ab062ab9630912c49/pkg/sql/sessiondata/internal.go#L93)):*
> ```Go
> 	// attempts to set a cluster setting overrode by the system tenant, the query
> 	// no-ops instead of errors.
> 	SwallowSettingOverrideErr bool
> ```

super nit: since we have so many different kinds of settings, this would be even more clear if it were named `SwallowClusterSettingOverrideErr`


<!-- Sent from Reviewable.io -->

msbutler commented:
closing in favor of https://github.com/cockroachdb/cockroach/pull/145332 which has a lighter impact.
